fs: uv_fs_{open,read,close}_dir

whitlockjc opened the pull request:
This PR is for taking the efforts by @misterdjules in #175 and applying those changes to `master` instead of `v1.x` per @saghul's suggestion.  You should be able to close #175 in favor of this but I'll leave that decision up to you.

**Note:** This commit also updates `uv__fs_readdir` to use `readdir` instead of the deprecated `readdir_r`.

whitlockjc commented:
cc'ing the required people...

/cc @saghul @piscisaureus @bnoordhuis

Fishrock123 commented:
Pinging @saghul @piscisaureus @bnoordhuis again :)

saghul commented:
Sorry, but I won't be able to review this anytime soon.

whitlockjc commented:
I have updated this PR for `master` deviation.  Test suite runs great, although `fs_event_watch_dir` and `fs_event_watch_dir_recursive` can randomly fail.  But that is unrelated to this PR and is fixed in #480.

rsms commented:
+1 — nice work.

eljefedelrodeodeljefe commented:
@whitlockjc I opened a related issue on the nodejs bug tracker. Without having a stake in this, can you rebase this, so it will be possible to review?

whitlockjc commented:
You got it.  I didn't realize this hadn't been merged.

eljefedelrodeodeljefe commented:
@saghul marked it v2 - maybe that was the reason. Not sure if it is though. If possible try not breaking ABI, then I guess it'll be easier to land. Would be really useful to land this, imo!

txdv commented:
It looks like its only adding additional API calls, how is it breaking anything?

eljefedelrodeodeljefe commented:
I am sure @saghul will clarify. Maybe it was just for roadmap.

whitlockjc commented:
It's been through a ton of review, there are tests and the ABI was not broken.  I'm not sure why it's marked as `v2` but it shouldn't be because of the ABI.  As for trying not to break the ABI, I don't think people do it unless they have to _(at least not after a full pass of code review)_, as I had to in #590.

whitlockjc commented:
@elijahandrews Do you mind linking this issue and the one in node?  That would be very useful.

eljefedelrodeodeljefe commented:
You mean me probably, right? It's about providing glob, where those APIs would be necessary nodejs/node#6677

whitlockjc commented:
Sorry about that, I see it was linked higher in the discussion but I didn't see it.

saghul commented:
This PR modifies the `uv_fs_t` structure: https://github.com/libuv/libuv/pull/416/files#diff-56343bd7ad8bf449c2693082fb3e4081R1107 That breaks the ABI, and the reason why I tagged it v2.

eljefedelrodeodeljefe commented on include/uv.h:
```diff
@@ -1096,6 +1104,9 @@ struct uv_fs_s {
   uv_stat_t statbuf;  /* Stores the result of uv_fs_stat() and uv_fs_fstat(). */
+  const uv_dir_t* dir; /* Stores the result of uv_fs_opendir() */
+  uv_dirent_t* dirents;
+  size_t nentries;
```
I guess it's possible without that addition, say with an own struct right? That of course breaks ABI compatibility.

addaleax replied:
Can you add methods similar to `uv_fs_get_statbuf` etc. from 0d6525acae3b9 ?

whitlockjc commented:
Doh...I stand corrected.  _(It's been quite a while since this was authored so I guess I forgot, and GitHub's diff viewer didn't help too much as I overlooked it.)_  Well, it all makes sense now.  Mark that up as two `v2` PRs from me that require ABI changes.  ;)

whitlockjc commented on include/uv.h:
```diff
@@ -1096,6 +1104,9 @@ struct uv_fs_s {
   uv_stat_t statbuf;  /* Stores the result of uv_fs_stat() and uv_fs_fstat(). */
+  const uv_dir_t* dir; /* Stores the result of uv_fs_opendir() */
+  uv_dirent_t* dirents;
+  size_t nentries;
```
Yup, I didn't see this.  _(I actually overlooked it twice because of the styling of the GitHub diff viewer...and this PR is pretty old so I just forgot.)_

saghul replied:
And where would you put those members or the new struct? FYI, I'll be -1 on bending over backwards for ABI compatibility in new features. That way we'd never start working on v2.

eljefedelrodeodeljefe commented:
That's fine. We'll get there. Give me a ping when I can support.

whitlockjc commented:
You got it.  I don't mind helping on the node side of things either if you need.  I'm already prepared for the node changes related to for #590 when the time comes.  :)

whitlockjc commented:
Since there is no immediate impact, I will punt on rebasing this since I've not heard anything about `v2` development just yet.  If anyone disagrees, let me know.

eljefedelrodeodeljefe commented:
@saghul I wouldn't mind if `v2` is imminent. What's the roadmap there?

saghul commented:
@eljefedelrodeodeljefe There is no defined roadmap as of yet. v2 won't happen overnight, that's for sure. Time is a scarce resource, I'm afraid. Personally I haven't had the time to work on any of the things I wanted for v2 yet.

saghul commented on include/uv-unix.h:
```diff
@@ -158,6 +158,13 @@ typedef gid_t uv_gid_t;
+ * "dirent" is used to hold a buffer large enough for any dirent in the
+ * directory being read. Avoids allocating for each directory entry.
+ */
+#define UV_DIR_PRIVATE_FIELDS \
```
style: align the backslashes

saghul commented on include/uv-win.h:
```diff
@@ -284,6 +284,29 @@ typedef struct uv__dirent_s {
+ * "dirent" is used to hold a buffer large enough for any dirent in the
+ * directory being read. It avoids allocating for each directory entry.
+ */
+#define UV_DIR_PRIVATE_FIELDS \
```
ditto

saghul commented on src/unix/fs.c:
```diff
@@ -120,6 +120,95 @@ static ssize_t uv__fs_fdatasync(uv_fs_t* req) {
 }
 
 
+static ssize_t uv__fs_clamp_max_path_len(ssize_t len) {
```
I believe we now have a similar thing in, which will conflict, can you please rebase and adjust?

saghul commented on src/unix/fs.c:
```diff
@@ -120,6 +120,95 @@ static ssize_t uv__fs_fdatasync(uv_fs_t* req) {
+#   if defined(NAME_MAX)
+      name_max = (NAME_MAX > 255) ? NAME_MAX : 255;
+#   else
+      return (size_t)(-1);
```
what does -1 mean?

whitlockjc replied:
I think the idea was that `-1` would indicate an error.  Maybe `0` would be better?

saghul replied:
I we can't be sure about the size I'd rather have a `#error` as we do for realpath. That would avoid potential buffer overruns, I think.

guymguym replied:
This entire function will not be needed with `readdir` instead of `readdir_r`, right?

saghul replied:
yep, I think so.

saghul commented on src/unix/fs.c:
```diff
@@ -120,6 +120,95 @@ static ssize_t uv__fs_fdatasync(uv_fs_t* req) {
+  return sizeof(struct dirent);
+}
+
+static ssize_t uv__fs_max_path_len(const char* path) {
```
same as above, I think you can remove this now

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+  * Allocate the space for each directory entry just once instead of
+  * once per directory entry.
+  */
+  dirent = malloc(len + 1);
```
use the libuv allocator: uv__malloc

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+   goto failed;
+  }
+
+  dir = malloc(sizeof(*dir));
```
ditto

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+  dir->dirent = dirent;
+
+  req->dir = dir;
+  req->ptr = (void*)req->dir;
```
unneeded cast I think

saghul replied:
also, why?

whitlockjc replied:
Without, you get this:

```
src/unix/fs.c:488:12: warning: assigning to 'void *' from 'const uv_dir_t *' (aka 'const struct uv_dir_s *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
  req->ptr = req->dir;
           ^ ~~~~~~~~
```

I'll see if there is a better approach.  _(I inherited some of this so bear with me)_

saghul replied:
sure!

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+failed:
+  if (errno == ENOMEM) {
+    if (dirent != NULL)
+      free(dirent);
```
uv__free

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+    if (dirent != NULL)
+      free(dirent);
+    if (dir != NULL)
+      free(dir);
```
ditto

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+
+    if (r == 0) {
+      if (res != NULL) {
+        dirents[dirent_idx].name = strdup(req->dir->dirent->d_name);
```
uv__strdup

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+  req->ptr = req->dirents;
+
+  for (dirent_idx = 0; dirent_idx < req->nentries; ++dirent_idx) {
+    r = readdir_r(req->dir->dir, req->dir->dirent, &res);
```
is it available in all platforms we support?

whitlockjc replied:
Looks to be based on some googling around.

guymguym replied:
@whitlockjc @saghul 
Hey, I see that `readdir_r` was deprecated in glibc by https://sourceware.org/ml/libc-alpha/2016-02/msg00093.html 
The reason is mostly about indecisive name length, see here - http://man7.org/linux/man-pages/man3/readdir_r.3.html
The suggestion is to use `readdir` which would also eliminate the need for `uv__fs_direntry_size` and its guesstimations.
In multithreaded environment, such as with uv fs workers, calling `readdir` on the same `DIR` stream is racy, and should be protected by mutex per dir.

saghul replied:
Wow, that is very usefult information, thanks a lot for pinting it out!

@whitlockjc we should probably add a mutex to out uv_dir_t struct, at least on Unix.

saghul commented on src/unix/fs.c:
```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+    }
+
+    if (req->dir->dirent) {
+      free(req->dir->dirent);
```
uv__free

saghul commented on src/uv-common.c:
```diff
@@ -496,37 +496,58 @@ int uv_fs_scandir_next(uv_fs_t* req, uv_dirent_t* ent) {
+    uv_dirent_t* dirents = req->ptr;
+    for (dirent_idx = 0; dirent_idx < req->nentries; ++dirent_idx) {
+      if (dirents[dirent_idx].name)
+        free((char*)dirents[dirent_idx].name);
```
unneded cast? also uv__free

whitlockjc replied:
As for the cast, like the one above it seems without this we end up with a compiler error:

```
src/unix/fs.c:1536:14: warning: passing 'const char *' to parameter of type 'void *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    uv__free(req->path);  /* Memory is shared with req->new_path. */
             ^~~~~~~~~
./src/uv-common.h:218:21: note: passing argument to parameter 'ptr' here
void uv__free(void* ptr);
```

saghul commented on src/win/fs.c:
```diff
@@ -1011,6 +1011,169 @@ void fs__scandir(uv_fs_t* req) {
+    return;
+  }
+
+  dir = malloc(sizeof(*dir));
```
uv__malloc

saghul commented on src/win/fs.c:
```diff
@@ -1011,6 +1011,169 @@ void fs__scandir(uv_fs_t* req) {
+    fmt = L"%s\\*";
+  }
+
+  path2 = (WCHAR*)malloc(sizeof(WCHAR) * (len + 4));
```
unneded cast, uv__malloc

whitlockjc replied:
Regarding the cast, same as the two instances above.

saghul commented on test/test-fs-readdir.c:
```diff
@@ -0,0 +1,514 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
please assign the copyright to "the libuv project"

whitlockjc replied:
I don't mind doing this but all other tests have the copyright currently being used in the PR.

saghul commented on test/test-fs-readdir.c:
```diff
@@ -0,0 +1,514 @@
+
+/* FIXME we shouldn't need to branch in this file */
+#if defined(__unix__) || defined(__POSIX__) || \
+	defined(__APPLE__) || defined(_AIX)
```
style: align these

whitlockjc replied:
Okay.

saghul commented on test/test-fs-readdir.c:
```diff
@@ -0,0 +1,514 @@
+  ASSERT(req == &readdir_req);
+  ASSERT(req->fs_type == UV_FS_READDIR);
+
+  // TODO jgilli: by default, uv_fs_readdir doesn't return 0 when reading
```
use /* */ for comments

saghul commented:
I started merging some things into master. Let's get this one in :-) Did a quick review round.

whitlockjc commented:
There are a ton of things to address, I will update the PR accordingly.

saghul commented:
@whitlockjc excellent, thanks! also, no rush.

whitlockjc commented:
I _think_ I addressed everything, and `make check` runs without issue.  Please let me know if you have more feedback that needs to be addressed.

saghul commented on include/uv-unix.h:
```diff
@@ -145,6 +145,13 @@ typedef gid_t uv_gid_t;
+ */
+#define UV_DIR_PRIVATE_FIELDS \
+  uv__dirent_t* dirent;       \
+  DIR* dir;
```
you might need a mutex here as readdir_r is deprecated...

saghul commented on include/uv.h:
```diff
@@ -1110,6 +1118,9 @@ struct uv_fs_s {
   void* ptr;
   const char* path;
   uv_stat_t statbuf;  /* Stores the result of uv_fs_stat() and uv_fs_fstat(). */
+  const uv_dir_t* dir; /* Stores the result of uv_fs_opendir() */
```
can you group these in an struct perhaps? That way accessing them would seem a bit more pleasant, I think.

saghul commented on src/unix/fs.c:
```diff
@@ -370,10 +437,10 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
     int i;
 
-    /* Memory was allocated using the system allocator, so use free() here. */
+    /* Memory was allocated using the system allocator, so use uv__free() here. */
```
does this comment still apply? maybe say "memory was allocated by libuv" ?

saghul commented on src/unix/fs.c:
```diff
@@ -370,10 +437,10 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
-      free(dents[i]);
-    free(dents);
+      uv__free(dents[i]);
+    uv__free(dents);
```
hold on, this change is on scandir, I think that was fine, because the library function allocates the memory, so we have to free it with `free` and not `uv__free`.

saghul commented on src/unix/fs.c:
```diff
@@ -382,6 +449,116 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
 }
 
+static int uv__fs_opendir(uv_fs_t* req) {
+  DIR *dir_stream;
```
style: DIR*

saghul commented on src/unix/fs.c:
```diff
@@ -382,6 +449,116 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+
+  len = uv__fs_direntry_size(dir_stream);
+  if (len == (size_t)-1)
+    return -1;
```
Hum, does this need to be a runtime check? If so, we'd need to set errno before returning -1.

whitlockjc replied:
I think this is just overly-cautious code and I'm not even sure what `errno` would be set to.  If you have an idea, I'll make it happen.

saghul commented on src/unix/fs.c:
```diff
@@ -382,6 +449,116 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+   goto failed;
+  }
+
+  dir = uv__malloc(sizeof(*dir));
```
these allocations will need to be adjusted when you switch to `readdir`

saghul commented:
@whitlockjc I started another round of review, but the change from `readdir_r` to `readdir` will change things a bit, so it's probably a better idea for me to hold off until then.

Basically, we need to have a mutex which needs to be acquired on `uv_fs_readdir` and `uv_fs_closedir` and after closing the dir handle it must be destroyed.

Sorry for not catching this earlier, I wasn't aware of the `readdir_r` deprecation.

whitlockjc commented:
Don't feel bad, it was deprecated until 2016/02, this PR predates that.  I'll get started.

cjihrig commented:
@whitlockjc is there any update here?

whitlockjc commented:
I dropped the ball.  I'll pick it back up.

DomVinyard commented:
@whitlockjc is there any update here? 😄

whitlockjc commented:
I apologize for dropping the ball again.  Company acquisition, vacation and then two months of holidays didn't help but it was ultimately my fault.

One thing that could speed this up is if we could [address the deprecations](https://github.com/libuv/libuv/pull/416#issuecomment-238304971) in a follow-up PR.  I'd hate to have to rewrite this so late in the game but I am cool with it.

I'll give @saghul a day or so to respond and if he doesn't, I'll assume we **must** address the deprecations and I'll get the PR addressed.

guymguym commented:
Hey, for what it's worth here's my opinion:
- In any case consider adding a mutex to the linux dir structure to avoid another breaking change.
- `uv__fs_direntry_size` with its compile-time & runtime cases is a kind of a baby black hole in terms of validation, and I would make an effort to either avoid it by making the change to use `readdir` or include tests for all these cases (tedious).

whitlockjc commented:
I'll start working on this.  I'll do my best to try to summarize the remaining changes to get a 👍 from the interested parties.  Feel free to chime in.

whitlockjc commented:
Updated to `master` to start with a clean slate.

rdeforest commented:
I posted a [comment](https://github.com/nodejs/node/issues/15699#issuecomment-336681982) on nodejs/node#15699 providing motivation for this work, should anyone be looking for some.

MylesBorins commented:
Hey all... just wanted to pop in and check on the status of this. I'd be willing to help move this forward

addaleax commented:
@MylesBorins If you* are interested in taking this up, rebasing & opening a new PR, I’d be happy to help solve issues that come up along the way

*or anybody else who’s reading this!

whitlockjc commented:
I'll gladly work on this as well.  I dropped the ball and since it wasn't on my daily radar, I've just not moved it along.  Can someone help me figure out what is required above/beyond the rebase?

addaleax commented:
@whitlockjc From what I’ve extracted from these threads, the other major thing was the usage of the deprecated `readdir_r()` function?

cjihrig commented:
@whitlockjc I'm hoping to do a v2 release in the next month or so. There is a merge of v1.x into master pending, and some other v2 PRs that can land.

If this lands on master, it would be able to go out in v2. Let's go through the rebase, review, CI run, etc. and see if we can get it landed.

whitlockjc commented:
I'll get started and do this in two pieces:

1. Get the PR with deprecated `readdir_r`
2. Update the PR without deprecated `readdir_r`

Seem reasonable?

sindresorhus commented:
> Seem reasonable?

@cjihrig ⬆️

guymguym commented:
Any help needed?

cjihrig commented:
Yes, that sounds reasonable. Sorry I missed your response.

whitlockjc commented:
Rebased against `master`.  I will now try to address @saghul's last round of changes.  Once that's done, I'll update for removing `readdir_r` usage.

whitlockjc commented:
It _might_ be easier to have a re-review but I leave that up to you.

whitlockjc commented:
Once we have this version of the PR reviewed, I'll move from `readdir_r` to `readdir`.  The code is simple but let's not lump multiple changes together.

whitlockjc commented:
PR updated `uv__fs_readdir` to use `readdir` instead of the depreated `readdir_r`.

guymguym commented on src/unix/fs.c:
```diff
@@ -421,6 +488,116 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+  if (dir_stream == NULL)
+    return -1;
+
+  len = uv__fs_direntry_size(dir_stream);
```
I see you still use `uv__fs_direntry_size` in order to malloc `dir->dirent` but it seems like it is unused now. Is that needed at all now that [readdir(3)](http://man7.org/linux/man-pages/man3/readdir.3.html) is returning the entries?

whitlockjc commented:
I _think_ I have addressed all feedback except the one from @addaleax asking for _"methods similar to uv_fs_get_statbuf etc. from 0d6525a"_.  Once the PR is good as-is, I'll add those changes in.  The reason I'm hesitant to do them is I had to go back and address some of @saghul's old feedback which was similar to @guymguym.  Once all of that stuff is solid, I'll address @addaleax's feedback.

guymguym commented:
@whitlockjc One last comment - I believe that for Linux you need to have a mutex per dir since `readdir()` is not thread safe and as far as I understand the libuv threadpool will be doing these readdir calls, right?

See what the [readdir(3)](http://man7.org/linux/man-pages/man3/readdir.3.html) docs has to say:
>In the current POSIX.1 specification (POSIX.1-2008), readdir() is not required to be thread-safe.  However, in modern implementations (including the glibc implementation), concurrent calls to readdir() that specify different directory streams are thread-safe.  In cases where multiple threads must read from the same directory stream, using readdir() with external synchronization is still preferable to the use of the deprecated readdir_r(3) function.  It is expected that a future version of POSIX.1 will require that readdir() be thread-safe when concurrently employed on different directory streams.

whitlockjc commented:
I thought something similar but figured it would come up in review.  Glad you brought it up.  I'll get it sorted once I've given enough soak time for further review.

bnoordhuis commented on test/test-fs-readdir.c:
```diff
@@ -0,0 +1,514 @@
+                  non_empty_readdir_cb);
+  }
+
+  uv_fs_req_cleanup(req);
```
This isn't safe, is it?  `req == &readdir_req` and that's (re)initialized on line 375.

bnoordhuis commented on test/test-fs-readdir.c:
```diff
@@ -0,0 +1,514 @@
+  ASSERT(r == 0);
+  uv_fs_req_cleanup(&close_req);
+
+    r = uv_fs_mkdir(loop, &mkdir_req, "test_dir/test_subdir", 0755, NULL);
```
Indent error.

whitlockjc commented:
I'll get on it.  I was waiting on #590 to be merged and forgot about this one.

rijnhard commented:
@whitlockjc ping, just making sure this hasn't slipped through the cracks

whitlockjc commented:
Thanks for the bump @rijnhard, I've just been super busy.  I'll make time for this by EOW.

rijnhard commented:
@whitlockjc ping

whitlockjc commented:
I'm back, sorry for the delay.  Work, life and other OSS.  :)

whitlockjc commented on src/unix/fs.c:
```diff
@@ -421,6 +421,91 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+  dir->dir = dir_stream;
+
+  req->dir = dir;
+  req->ptr = (void*)req->dir;
```
Compiler warning when I remove the cast:

```
src/unix/fs.c:499:12: warning: assigning to 'void *' from 'const uv_dir_t *' (aka 'const struct uv_dir_s *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
  req->ptr = req->dir;
           ^ ~~~~~~~~
1 warning generated.
  CCLD     libuv.la
```

whitlockjc commented on src/unix/fs.c:
```diff
@@ -421,6 +421,91 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
+      } else {
+        return dirent_idx;
+      }
+    }
```
What should we do here?

whitlockjc commented:
Would anyone else like to pick this up and drive it?  The review feedback from @guymguym and @bnoordhuis is going to take more time than I can seem to come up with right now.

cjihrig commented:
> Would anyone else like to pick this up and drive it?

I'll pick this up.

guymguym commented:
@cjihrig Let me know if I can help too

Fishrock123 commented:
Do we know where `master` is going yet? If someone were to update this, is `master` still a reasonable target, or will that never make it into a release?

rijnhard commented:
This died again didn't it

cjihrig commented:
It hasn't died again. I promise.

cjihrig commented:
Moving to #2057 with most of @bnoordhuis nits addressed from the last review, as well as some comments of my own.
